src/caip-types.ts

```diff
@@ -0,0 +1,165 @@
+import type { Infer } from 'superstruct';
+import {
+  array,
+  is,
+  object,
+  optional,
+  pattern,
+  size,
+  string,
+} from 'superstruct';
+
+export const CHAIN_ID_REGEX =
+  /^(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-a-zA-Z0-9]{1,32})$/u;
+
+export const ACCOUNT_ID_REGEX =
+  /^(?<chainId>(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-a-zA-Z0-9]{1,32})):(?<accountAddress>[a-zA-Z0-9]{1,64})$/u;
+
+/**
+ * Parse a chain ID string to an object containing the namespace and reference.
+ * This validates the chain ID before parsing it.
+ *
+ * @param caipChainId - The chain ID to validate and parse.
+ * @returns The parsed chain ID.
+ */
+export function parseCaipChainId(caipChainId: CaipChainId): {
+  namespace: CaipNamespaceId;
+  reference: string;
+} {
+  const match = CHAIN_ID_REGEX.exec(caipChainId);
+  if (!match?.groups) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  return {
+    namespace: match.groups.namespace as CaipNamespaceId,
+    reference: match.groups.reference as string,
+  };
+}
+
+/**
+ * Parse an account ID to an object containing the chain, chain ID and address.
+ * This validates the account ID before parsing it.
+ *
+ * @param accountId - The account ID to validate and parse.
+ * @returns The parsed account ID.
+ */
+export function parseCaipAccountId(accountId: CaipAccountId): {
+  chain: { namespace: CaipNamespaceId; reference: string };
+  chainId: CaipChainId;
+  address: string;
+} {
+  const match = ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress as string,
+    chainId: match.groups.chainId as CaipChainId,
+    chain: {
+      namespace: match.groups.namespace as CaipNamespaceId,
+      reference: match.groups.reference as string,
+    },
+  };
+}
+
+/**
+ * A helper struct for a string with a minimum length of 1 and a maximum length
+ * of 40.
+ */
+export const LimitedString = size(string(), 1, 40);
+
+/**
+ * A CAIP-2 chain ID, i.e., a human-readable namespace and reference.
+ */
+export const CaipChainIdStruct = pattern(string(), CHAIN_ID_REGEX);
+export type CaipChainId = `${string}:${string}`;
```

Comment:
```suggestion
export type CaipChainId = Infer<typeof CaipChainIdStruct>;
```
This is a change I'm making in my name resolver PR, I think this type is better.

Reply:
This loosens the type to just `string` though. I'm not sure this is what we want

Reply:
Oh! I didn't realize it was loosening the type. Ok to keep as is then.